[quant][graphmode] quantization support for prim::CallFunction

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35385 [quant][graphmode] Remove unused patterns
* #35347 [quant][graphmode] Quantization support for `aten::conv3d`
* #35345 [quant][graphmode][fix] Add filter for `quantized::add`
* #35336 [quant][graphmode] qconfig_dict support None
* #35420 [quant][graphmode] Make `aten::relu` a general op
* #35335 [quant][graphmode] `prepare_script` takes original qconfig_dict
* #35334 [quant][graphmode] Quantization support for `quantized::add_scalar`
* #35333 [quant][graphmode][refactor] Support filter function in quant fusion patterns
* #35332 [quant][graphmode] Quantization support for permute and repeat_interleave
* #35130 [quant][graphmode] Make interpolate/upsample work again
* #35142 [quant][graphmode] SwapDeQuant support prim::If
* #35141 [quant][graphmode][refactor] getGeneralOpTensorInputIndexes -> getGeneralOpTensorInputs
* #35135 [quant][graphmode][refactor] swapDeQuant takes block as arugment
* #35077 [quant][graphmode] Fold quantized prepacking ops
* #35073 [quant] Make conv2d_prepack and linear_prepack pure
* **#34855 [quant][graphmode] quantization support for prim::CallFunction**

Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20655305](https://our.internmc.facebook.com/intern/diff/D20655305)

jerryzh168 committed:
[quant][graphmode] quantization support for prim::CallFunction

Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] propagate observed property for user defined prim::CallFunction"


Summary:
This only supports the case when user defined function does not require inserting new observers,
it will error out if we need to insert observers for a user defined function. Reason is function does not have states and unlike methods we don't have a handle to an object that is stateful, so we don't support that right now.

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -64,6 +64,18 @@ struct PatternsAndModules {
   Module packed_params_module;
 };
 
+std::vector<std::string> _single_input_general_call_funcs = {
```
Can you add a comment to explain what these mean? Is the first list functions that do dispatch for qtensor under the hood, so that you can propagate dq?

jerryzh168 replied:
will do

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -194,6 +199,21 @@ bool nodeQuantizable(Node* n) {
     });
 }
 
+bool userDefinedCallFunction(Node* n) {
```
Why is this called user defined? Isnt this a function that is not one of the functions listed in the arguments of the call? This can be a aten function isnt it?

jerryzh168 replied:
there are "builtin" CallFunction that we don't want to go through their graph and do quantization because we want to preserve the boundary, e.g. `linear`

raghuramank100 reviewed (changes requested):
Requesting some clarification before you land it.

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] quantization support for prim::CallFunction"


Summary:

Test Plan:
python test/test_jit.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20655305](https://our.internmc.facebook.com/intern/diff/D20655305)

[ghstack-poisoned]
